### PR TITLE
Support-Multi-version-TiDB-monitor

### PR DESCRIPTION
**Pull request by qiffang**
### What problem does this PR solve? 
In this PR, it contains 3 functions
1. Support multi TiDB version
2. Support to dynamic reload rules
3. Update monitor deployment Strategy

1.TiDB multi version
In this [PR](https://github.com/pingcap/monitoring), Every TiDB has own monitor initializer image.
And the docker receive 4 variables
```
GF_PROVISIONING_PATH // grafana provisioning path
TIDB_CLUSTER_NAME // TiDB cluster name
TIDB_ENABLE_BINLOG // whether enable binlog
PROM_CONFIG_PATH // proemtheus rules config path
```
TiDB Operator use the new monitor way to replace the old.

2.Dynamic update alert rules
In [reload](https://github.com/pingcap/monitoring/tree/master/reload), it support a UI to update alert rules and it will call reload of Promtheus.
The UI likes this
image

3. Update strategy with ReCreate.
### What is changed and how does it work?
Delete the old way and import a tidb-monitor-initializer image to init grafana dashboards and prometheus rules.

### Check List 

Tests 

 - TiDB Cluster dashboards work well and ClusterName is replaced successful.
 - Promethes rules work well and ClusterName is replaced successful.
-  Alert rule can be updated dynamic.
-  Monitor deployment strategy is updated successful.

Related changes

 - Need to update the documentation

### Does this PR introduce a user-facing change?:
 
 ```release-note
Action required: Users should migrate the configs in values.yaml of previous chart releases to the new values.yaml of the new chart. Otherwise, the monitor pods will be failed when upgrading monitor with the new chart.
For example, configs in old values.yaml:
monitor:
  create: true
  # Also see rbac.create
  # If you set rbac.create to false, you need to provide a value here.
  # If you set rbac.create to true, you should leave this empty.
  # serviceAccount:
  persistent: false
  storageClassName: local-storage
  storage: 10Gi
  grafana:
    ...
  prometheus:
    image: prom/prometheus:v2.2.1
    ...

After migration, new values.yaml should be as below:
monitor:
  create: true
  # Also see rbac.create
  # If you set rbac.create to false, you need to provide a value here.
  # If you set rbac.create to true, you should leave this empty.
  # serviceAccount:
  persistent: false
  storageClassName: local-storage
  storage: 10Gi
  initializer:
    image: pingcap/tidb-monitor-initializer:v3.0.0
    imagePullPolicy: IfNotPresent
    reloader:
    create: true
    image: pingcap/tidb-monitor-reloader:v1.0.0
    imagePullPolicy: IfNotPresent
    service:
      type: NodePort
  grafana:
    ...
  prometheus:
    image: prom/prometheus:v2.11.1
    ...
```

### CODE REVIEW

File: charts/tidb-cluster/values.yaml [hunk 1]
```diff
@@ -330,6 +330,10 @@ monitor:
   persistent: false
   storageClassName: local-storage
   storage: 10Gi
+  initializer:
+    create: true
```
Comment (weekface) [inline charts/tidb-cluster/values.yaml]: delete this attribute.

File: charts/tidb-cluster/values.yaml [hunk 2]
```diff
@@ -330,6 +330,10 @@ monitor:
...
+    image: pingcap/tidb-monitor-initializer:v3.0.0-rc.1
```
Comment (weekface) [inline charts/tidb-cluster/values.yaml]:
```
$ docker pull pingcap/tidb-monitor-initializer:v3.0.0-rc.1
Error response from daemon: pull access denied for pingcap/tidb-monitor-initializer, repository does not exist or may require 'docker login'
```

Comment (qiffang) [reply charts/tidb-cluster/values.yaml]: https://github.com/pingcap/monitoring 和 这个PR 是同一个ticket，这个PR里面的镜像还没上传，我是本地测的，正在和强哥沟通。

File: charts/tidb-cluster/values.yaml (hunk 1 above)
Comment (qiffang) [reply charts/tidb-cluster/values.yaml]: Removed

File: charts/tidb-cluster/values.yaml (hunk 2 above)
Comment (tennix) [reply charts/tidb-cluster/values.yaml]: The values in other places like test, terraform also have to be updated.

File: charts/tidb-cluster/templates/monitor-deployment.yaml [hunk 3]
```diff
@@ -181,23 +185,10 @@ spec:
           - key: dashboard-config
             path: dashboards.yaml
         name: dashboards-provisioning
+      - emptyDir: {}
```
Comment (tennix) [inline charts/tidb-cluster/templates/monitor-deployment.yaml]: Is it better to use the same directory as the data directory?

File: charts/tidb-cluster/values.yaml (hunk 2 above)
Comment (gregwebs) [reply charts/tidb-cluster/values.yaml]: We could have a CI check to tell us that terraform values are out of date.

File: charts/tidb-cluster/templates/monitor-service.yaml [hunk 4]
```diff
@@ -27,6 +27,30 @@ spec:
     app.kubernetes.io/component: monitor
 {{- end }}
 ---
+{{- if .Values.monitor.reloader.create }}
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ template "cluster.name" . }}-reloader
```
Comment (tennix) [inline charts/tidb-cluster/templates/monitor-service.yaml]: Should add `monitor` in the service name.

File: charts/tidb-cluster/templates/monitor-deployment.yaml [hunk 5]
```diff
@@ -61,6 +63,25 @@ spec:
         volumeMounts:
         - name: monitor-data
           mountPath: /data
+      - name: monitor-initializer
```
Comment (tennix) [inline charts/tidb-cluster/templates/monitor-deployment.yaml]: I think it's better to merge the two init containers into one.

File: charts/tidb-cluster/templates/monitor-deployment.yaml [hunk 6]
```diff
@@ -11,6 +11,9 @@ metadata:
     helm.sh/chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+"  "_" }}
 spec:
   replicas: 1
+  strategy:
+    type: Recreate
+    rollingUpdate: null
```
Comment (weekface) [inline charts/tidb-cluster/templates/monitor-deployment.yaml]: remove this?

Comment (qiffang) [reply charts/tidb-cluster/templates/monitor-deployment.yaml]: remove这个会有问题
```
Error: UPGRADE FAILED: Deployment.apps “tikv200-monitor” is invalid: spec.strategy.rollingUpdate: Forbidden: may not be specified when strategy `type` is ‘Recreate’
```
需要显示把这个置为null

Comment (weekface) [reply charts/tidb-cluster/templates/monitor-deployment.yaml]: ok

File: charts/tidb-cluster/templates/monitor-deployment.yaml (hunk 3 above)
Comment (qiffang) [reply charts/tidb-cluster/templates/monitor-deployment.yaml]: 数据目录是 PV，配置不需要PV

File: charts/tidb-cluster/templates/monitor-service.yaml (hunk 4 above)
Comment (qiffang) [reply charts/tidb-cluster/templates/monitor-service.yaml]: Done

File: charts/tidb-cluster/templates/monitor-deployment.yaml (hunk 5 above)
Comment (qiffang) [reply charts/tidb-cluster/templates/monitor-deployment.yaml]: Done

File: charts/tidb-cluster/templates/monitor-deployment.yaml [hunk 7]
```diff
@@ -47,18 +50,34 @@ spec:
...
-        imagePullPolicy: {{ .Values.monitor.prometheus.imagePullPolicy | default "IfNotPresent" }}
+      - name: monitor-initializer
+        image: {{ .Values.monitor.initializer.image }}
+        imagePullPolicy: {{ .Values.monitor.initializer.imagePullPolicy | default "IfNotPresent" }}
+        env:
+        - name: GF_PROVISIONING_PATH
+          value: /grafana-dashboard-definitions/tidb
+        - name: TIDB_CLUSTER_NAME
+          value: {{ template "cluster.name" . }}
+        - name: TIDB_ENABLE_BINLOG
```
Comment (weekface) [inline charts/tidb-cluster/templates/monitor-deployment.yaml]: why add this env? `TIDB_ENABLE_BINLOG `

File: charts/tidb-cluster/values.yaml [hunk 8]
```diff
@@ -336,6 +336,15 @@ monitor:
   persistent: false
   storageClassName: local-storage
   storage: 10Gi
+  initializer:
+    image: pingcap/tidb-monitor-initializer:v3.0.0
```
Comment (tennix) [inline charts/tidb-cluster/values.yaml]: This should be v3.0.1 to match the default tidb version.

Comment (qiffang) [reply charts/tidb-cluster/values.yaml]: Done

File: charts/tidb-cluster/templates/monitor-deployment.yaml (hunk 7 above)
Comment (qiffang) [reply charts/tidb-cluster/templates/monitor-deployment.yaml]: This env is used to decide whether Grafana show BinLog dashboard.

**Pull request merged**